File: leetcode/94-binary-tree-inorder-traversal.py
# https://leetcode.com/problems/binary-tree-inorder-traversal/ (top-interview-questions)

from typing import List, Optional


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
        
class Solution:
    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        stack, values = list(), list()
        while stack or root:
            if root:
                stack.append(root)
                root = root.left
            else:
                root = stack.pop()
                values.append(root.val)
                root = root.right
        return values
    
    def inorder_traversal(self, root: Optional[TreeNode]) -> List[int]:
        stack, values = list(), list()
        while stack or root:
            while root:
                stack.append(root)
                root = root.left
            root = stack.pop()
            values.append(root.val)
            root = root.right
        return values
    

File: leetcode/94-binary-tree-inorder-traversal.js
// https://leetcode.com/problems/binary-tree-inorder-traversal/ (top-interview-questions)


// Definition for a binary tree node.
function TreeNode(val, left, right) {
 this.val = (val===undefined ? 0 : val)
 this.left = (left===undefined ? null : left)
 this.right = (right===undefined ? null : right)
}

/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var inorderTraversal = function(root) {
  
  const values = [];
  const stack = [];

  while (stack.length > 0 || root) {
    if (root) {
      stack.push(root);
      root = root.left;
    } else {
      root = stack.pop();
      values.push(root.val);
      root = root.right;
    }
  }

  return values;

};
